src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs

```diff
@@ -0,0 +1,403 @@
+                                args.cache[tr] = transition;
+                            }
+                            break;
+
+                        default:
+                            Debug.Assert(tr._first is not null && tr._second is not null);
+
+                            // Push the tr for the second time
+                            work.Push((tr, true));
+
+                            // Push the branches also, unless they have been computed already
+                            if (!args.cache.ContainsKey(tr._second))
+                            {
+                                work.Push((tr._second, false));
+                            }
+
+                            if (!args.cache.ContainsKey(tr._first))
+                            {
+                                work.Push((tr._first, false));
+                            }
+
+                            break;
+                    }
+                }
+            }
+
+            return args.cache[tregex];
+        }
+
+        /// <summary>Representation of transitions inside the parent class</summary>
+        private class Transition
+        {
+            public readonly TransitionRegexKind _kind;
+            public readonly int _leaf;
+            public readonly S? _test;
+            public readonly SymbolicRegexNode<S>? _look;
+            public readonly Transition? _first;
+            public readonly Transition? _second;
+
+            public static readonly Transition s_deadend = new Transition(TransitionRegexKind.Leaf, leaf: s_deadend_state);
+            public static readonly Transition s_unexplored = new Transition(TransitionRegexKind.Leaf, leaf: s_unexplored_state);
+
+            internal Transition(TransitionRegexKind kind, int leaf = 0, S? test = default(S), SymbolicRegexNode<S>? look = null, Transition? first = null, Transition? second = null)
+            {
+                _kind = kind;
+                _leaf = leaf;
+                _test = test;
+                _look = look;
+                _first = first;
+                _second = second;
+            }
+
+            /// <summary>Enumerates all the paths in this transition excluding paths to dead-ends (and unexplored states if any)</summary>
+            internal IEnumerable<(S, SymbolicRegexNode<S>?, int)> EnumeratePaths(IBooleanAlgebra<S> solver, S pathCondition)
+            {
+                switch (_kind)
+                {
+                    case TransitionRegexKind.Leaf:
+                        // Omit any path that leads to a deadend or is unexplored
+                        if (_leaf >= 0)
+                        {
+                            yield return (pathCondition, null, _leaf);
+                        }
+                        break;
+
+                    case TransitionRegexKind.Union:
+                        Debug.Assert(_first is not null && _second is not null);
+                        foreach (var path in _first.EnumeratePaths(solver, pathCondition))
```

Comment:
Nit: for all of these `var`s, can we use the actual type?

Reply:
ok, will do